x/storage/keeper/payment.go

```diff
@@ -309,12 +346,43 @@ func (k Keeper) ChargeViaBucketChange(ctx sdk.Context, bucketInfo *storagetypes.
 		return fmt.Errorf("get new bucket bill failed: %s %w", bucketInfo.BucketName, err)
 	}
 
-	// charge according to bill change
-	err = k.ApplyBillChanges(ctx, prevBill, newBill)
-	if err != nil {
-		ctx.Logger().Error("charge via bucket change failed", "bucket", bucketInfo.BucketName, "err", err.Error())
-		return err
+	if ctx.IsUpgraded(upgradetypes.Serengeti) {
+		isPreviousBucketLimited := k.isBucketRateLimited(ctx, bucketInfo.BucketName)
+
+		if prevPaymentAccount == bucketInfo.PaymentAddress && isPreviousBucketLimited {
+			// todo: do we need to contraint this
+			return fmt.Errorf("payment account is not changed but the bucket is limited")
+		}
+
+		// check the bucket's new flow rate limit
+		err := k.isBucketFlowRateUnderLimit(ctx, sdk.MustAccAddressFromHex(bucketInfo.PaymentAddress), sdk.MustAccAddressFromHex(bucketInfo.Owner), bucketInfo.BucketName, newBill)
+		if err != nil {
+			return err
+		}
+
+		if isPreviousBucketLimited {
+			err = k.ApplyBillChanges(ctx, nil, &newBill)
```

Comment:
This seems dead code, never will be executed

Reply:
it will be executed if the bucket is rate limited before